### PR TITLE
(draft) Defer File closing to a background thread pool and cache dire…

### PR DESCRIPTION
**Pull request by jpdoyle**
…ctory IDs

My current approach to #64

### CODE REVIEW

File: sources/file_table.cpp [hunk 1]
```diff
@@ -175,6 +177,15 @@ FileTable::~FileTable()
 FileBase* FileTable::open_as(const id_type& id, int type)
 {
     auto it = m_files.find(id);
+    if(it == m_files.end()) {
+        std::lock_guard<std::mutex> l(m_closing_lock);
+        auto close_it = m_files_to_close.find(id);
+        if(close_it != m_files_to_close.end()) {
+            m_files.emplace(id,std::move(close_it->second));
+            m_files_to_close.erase(close_it);
+            it = m_files.find(id);
```
Comment (netheril96) [inline sources/file_table.cpp]: `emplace` returns the iterator to the created element. Another `find` is unnecessary time cost.

File: sources/file_table.cpp [hunk 2]
```diff
@@ -191,6 +202,20 @@ FileBase* FileTable::open_as(const id_type& id, int type)
...
+            {
+                std::lock_guard<std::mutex> l(m_closing_lock);
+                being_closed = (m_closing_ids.count(id) != 0);
+            }
+            if(being_closed) {
+                free_pool.wait();
+            }
+        } while(being_closed);
+    }
+
```
Comment (netheril96) [inline sources/file_table.cpp]: What is the purpose of this block of code?

File: sources/file_table.cpp [hunk 3]
```diff
@@ -264,11 +289,49 @@ void FileTable::close(FileBase* fb)
...
+            std::function<void()> free_ptr = [id,this]() {
+                FileBase* ptr = nullptr;
+                {
+                    std::lock_guard<std::mutex> l(m_closing_lock);
+                    auto it = m_files_to_close.find(id);
+                    if(it == m_files_to_close.end()) {
+                        // The file was reopened
+                        return;
+                    }
+                    ptr = it->second.release();
```
Comment (netheril96) [inline sources/file_table.cpp]: The two lines below may throw exceptions, causing leakage of `ptr`. Instead of raw pointer, it should be a `std::unique_ptr`.

File: sources/file_table.cpp (hunk 2 above)
Comment (jpdoyle) [reply sources/file_table.cpp]: This is to make sure that the file is not currently being deleted by one of the background threads -- I don't know the FileBase implementation well enough to know that it would be safe to have two FileBases referring to the same file open, with one actively being deleted by a background thread. 

In pseudocode, it roughly does:

    while(the file I'm trying to open is being closed) {
        wait for at least one background thread to finish (which would indicate a change to the list of files being closed)
    }

File: sources/thread_pool.hpp [hunk 4]
```diff
@@ -0,0 +1,123 @@
+#ifndef THREAD_POOL_HPP
```
[the remark this answers is not in the document]
Comment (jpdoyle) [reply sources/thread_pool.hpp]: 5000 threads is definitely overkill, but I haven't tuned that number yet. In theory, running background threads might not be necessary at all, but I haven't dug into what deleting a `FileBase` exactly does -- it seems to be tied up in IO calls, but I'm not sure which ones (careful use of `strace` might be informative). With my folder I'm seeing a ~50x speedup -- I had timed `ls` before, but `find -type f` is more representative -- with this change it takes 2 minutes 11 seconds, and with vanilla `securefs` it took 1 hour 51 minutes. For reference, `find -type f | wc -l` says that this directory has 250289 files.

**Comment by jpdoyle** [comment]
I think the travis OSX failure was because the thread pool spawned 5000 threads. I reduced it to 50, which should be more reasonable.

File: sources/thread_pool.hpp [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        /* std::cout << "Adding to queue " << i << std::endl; */
+        {
+            std::lock_guard<std::mutex> l(*locks[i]);
+
+            work_queues[i].push(j);
+        }
+        condvars[i]->notify_all();
+    }
+
+    bool done() {
```
Comment (netheril96) [inline sources/thread_pool.hpp]: I have a hard time following what `done()` and `wait()` does, or what is the use of `condvars`. Besides, `done()` seems susceptible to race condition, as it accesses `work_queues` without acquiring the corresponding lock.

Comment (jpdoyle) [reply sources/thread_pool.hpp]: `condvars` lets the worker threads sleep without wasting any CPU time until there is work available for them to do -- `condvars[i]->wait(l)` pauses the thread until `convars[i]->notify_all()` is called on some other thread -- while the thread is asleep, the lock `l` is unlocked, but once it wakes up again it aquires `l` once more.

`done()` has a minor data race, but it's a read-only operation. Between jobs being added with `add_job` and finishing, `done()` will never `true`. If there are many jobs being added and finishing simultaneously, `done` might erroneously return `true`. `wait` waits for some job to complete, or for 100ms to pass, whichever happens first.

I should add these as comments :)

**Comment by jpdoyle** [comment]
I agree about the cache -- I'm gonna run a small benchmarking matrix using `ls` and `find -type f` to double-check how much of an effect the caching has, but I think my implementation is too half-baked to ship. A true solution would need to actively monitor the underlying folder with something like `inotify`, and even that isn't perfectly reliable.

File: sources/thread_pool.hpp (hunk 5 above)
Comment (netheril96) [reply sources/thread_pool.hpp]: Race condition invokes undefined behavior. Nothing, not even reading, is safe during a race.

But more importantly, I don't think we should have `done()` and `wait()`. I guess that the intention is signal to the main threading calling `open_as` when to wake up and recheck if the ID to be opened is no longer being closed. But

* Opening a ID that is just being closed is not a common scenario.
* The signal is not accurate. It only signals that one ID has been closed, not which ID.
* Each background thread has to call `job_done.notify_all()` after each closing. Not only does each call cost CPU cycles, it reduces parallelism since several threads may call it at the same time and have to wait for others.

In a nutshell, this design incurs a great cost, for a not-so-common scenario, and the main thread still has to recheck many times. Too much cost, too little gain.

Comment (jpdoyle) [reply sources/thread_pool.hpp]: Good catch on the undefined behavior. I've been doing a bit too much platform-locked stuff so I forgot the CPP generic memory model. Even platform-locked to x86 this code may sometimes fail due to memory reordering effects in `empty()` -- so the body of the `if` should be in a lock guard regardless.

> Opening an ID that is just being closed is not a common scenario

Many directories in `find -type f` will fit this profile, with the current length of time deleting a `FileBase` takes.

> The signal is not accurate. It only signals that one ID has been closed, not which ID.

This is correct, but the signal *will* always trigger when the correct ID is closed, so it's just the pull-based version of this communication protocol. To wait on a particular ID, we'd have to have the thread pool know a lot more about the jobs being given to it. That *might* be a good idea, but given how the data race in `done()` snuck in, I think adding code complexity to the thread pool is risky.

> Each background thread has to call job_done.notify_all() after each closing. Not only does each call cost CPU cycles, it reduces parallelism since several threads may call it at the same time and have to wait for others.

Using something like condition variables will waste a lot less CPU time than periodic polling, and when an extended wait isn't necessary the condition variable will return quickly.

Comment (netheril96) [reply sources/thread_pool.hpp]: It is best if you could do a real world benchmark on which approach is better (and how many threads are appropriate). If you insist on current design, though, I won't object.

File: sources/file_table.cpp [hunk 6]
```diff
@@ -191,6 +205,20 @@ FileBase* FileTable::open_as(const id_type& id, int type)
         }
     }
 
+    if(!free_pool.done()) {
+        bool being_closed = false;
+        do {
+            being_closed = false;
```
Comment (netheril96) [inline sources/file_table.cpp]: This is a no-op.
